### ISSUE TITLE
MessageMeta trait for Any::type_url

### ISSUE THREAD

**Issue by fdeantoni** [danburkert/prost#247]
As far as I understand Prost does not store any details about the message itself, such as the message name that was used to generate it, and the package name that was defined in the proto file. This information can be useful though. For example, when "packing" a message into an Any type, the `type_url` usually takes the form of `type.googleapis.com/package_name.message_name`. 

For example, in one of our java projects, we have a proto file as follows:
```protobuf
syntax = "proto3";

import "google/protobuf/any.proto";

package my.project;

message ServerRequest {
    string requestId = 1;
    google.protobuf.Any payload = 2;
}

message Query {
    string key = 1;
}
```

When using the [Java Any API](https://developers.google.com/protocol-buffers/docs/reference/java), when you "pack" the Query message into an Any, it will generate a type_url as `type.googleapis.com/my.project.Query`.

To support a similar "pack" API as the Java, C++ and Python implementations have, perhaps each Message struct can implement a MessageMeta trait with the following:
```rust
pub trait MessageMeta {
    fn name(&self) -> &'static str;
    fn package_name(&self) -> &'static str;
}
```

The "pack" and "unpack" API could then look something as follows:
```rust
#[test]
fn test_well_known_types_any() {
    let msg = Foo {
        null: ::prost_types::NullValue::NullValue.into(),
        timestamp: Some(::prost_types::Timestamp {
            seconds: 99,
            nanos: 42,
        }),
    };
    let any = Any::pack(msg);
    println!("{:?}", any);
    let unpacked = any.unpack(Foo::default()).unwrap();
    println!("{:?}", unpacked);
}
```

I got this working in my cloned repo (see branch [pb-any](https://github.com/fdeantoni/prost/tree/pb-any)) and so far it seems to work ok. Downside of course is that each message now implements a trait called MessageMeta, but I suppose this can be turned on/off by defining it as a feature.

**Comment by fdeantoni** [comment]
Addressed in pull request .#255

**Comment by fdeantoni** [comment]
Closed in favour of #277.
